### PR TITLE
feat(query-orchestrator): Introduce CubeStoreQueueDriver

### PR DESCRIPTION
**Pull request by ovr**
Hello!

It's an initial version of the cube store queue driver.

refs https://github.com/cube-js/cube.js/issues/5042

### General queue:

```mermaid
sequenceDiagram
    participant BackgroundQueryQueue
    participant QueueQueue
    participant QueueDriverInterface
    participant CubeStore

    QueueQueue->>QueueDriverInterface: getResult
    QueueDriverInterface->>+CubeStore: QUEUE RESULT ?
    QueueDriverInterface-->>+QueueQueue: QueryResult|null
    deactivate CubeStore

    QueueQueue->>QueueDriverInterface: addToQueue
    QueueDriverInterface->>+CubeStore: QUEUE ADD PRIORITY N key payload
    QueueDriverInterface-->>+QueueQueue: AddToQueueResponse

    loop reconcileQueueImpl
        QueueQueue->>QueueDriverInterface: getQueriesToCancel
        QueueQueue->>QueueDriverInterface: getQueryAndRemove
        QueueDriverInterface->>CubeStore: QUEUE TO_CANCEL ?stalled_timeout ?orphaned_timeout ?prefix

        QueueQueue->>QueueDriverInterface: getActiveQueries
        QueueDriverInterface->>CubeStore: QUEUE ACTIVE ?prefix
        QueueDriverInterface-->>+QueueQueue: getActiveQueriesResponse

        QueueQueue->>QueueDriverInterface: getToProcessQueries
        QueueDriverInterface->>CubeStore: QUEUE PENDING ?prefix
        QueueDriverInterface-->>+QueueQueue: getToProcessQueriesResponse

        QueueQueue-)+BackgroundQueryQueue: processQuery
        Note over QueueQueue,BackgroundQueryQueue: Async call to procesQuery, which doesnt block here
    end

    alt lookUpInActive: Lookup query in processing
        QueueQueue->>QueueDriverInterface: getQueryDef
        activate CubeStore
        QueueDriverInterface->>CubeStore: QUEUE GET ?key
        CubeStore-->>+QueueQueue: QueryDef|null
        deactivate CubeStore

        QueueQueue->>QueueDriverInterface: getQueryStageState
        activate CubeStore
        QueueDriverInterface->>CubeStore: QUEUE LIST
        CubeStore-->>+QueueQueue: TODO
        deactivate CubeStore
        Note over QueueQueue,QueueDriverInterface: Show waiting for query
    end

    QueueQueue->>QueueDriverInterface: getResultBlocking
    activate CubeStore
    QueueDriverInterface->>CubeStore: QUEUE RESULT_BLOCKING ?timeout ?key
    CubeStore-->>+QueueQueue: QueryResult|null
    deactivate CubeStore
```

### Background execution process:

```mermaid
sequenceDiagram
    participant QueryOrchestrator
    participant BackgroundQueryQueue
    participant QueueDriverInterface
    participant CubeStore

    loop processQuery: Background execution
        BackgroundQueryQueue->>QueueDriverInterface: getNextProcessingId
        activate CubeStore
        QueueDriverInterface->>CubeStore: CACHE INCR ?
        CubeStore-->>+BackgroundQueryQueue: number
        deactivate CubeStore

        BackgroundQueryQueue->>QueueDriverInterface: retrieveForProcessing
        activate CubeStore
        QueueDriverInterface->>CubeStore: QUEUE RETRIEVE CONCURRENCY ?number ?key
        CubeStore-->>+BackgroundQueryQueue: QueryDef
        deactivate CubeStore

        BackgroundQueryQueue->>QueueDriverInterface: optimisticQueryUpdate
        activate CubeStore
        QueueDriverInterface->>CubeStore: QUEUE MERGE_EXTRA ?key {"startTime"}
        CubeStore-->>+BackgroundQueryQueue: ok
        deactivate CubeStore

        BackgroundQueryQueue->>QueueDriverInterface: optimisticQueryUpdate
        activate CubeStore
        QueueDriverInterface->>CubeStore: QUEUE MERGE_EXTRA ?key {"cancelHandler"}
        CubeStore-->>+BackgroundQueryQueue: ok
        deactivate CubeStore

        par executing: Query
            BackgroundQueryQueue->>QueueDriverInterface: updateHeartBeat
            QueueDriverInterface-->>BackgroundQueryQueue: ok
            Note over BackgroundQueryQueue,QueueDriverInterface: intervalTimer

            BackgroundQueryQueue->>QueryOrchestrator: execute
            QueryOrchestrator-->>BackgroundQueryQueue: result
        end

        BackgroundQueryQueue->>QueueDriverInterface: setResultAndRemoveQuery
        activate CubeStore
        QueueDriverInterface->>CubeStore: QUEUE ACK ?key ?result
        CubeStore-->>+BackgroundQueryQueue: ok
        deactivate CubeStore
    end
```

### Testing

image

### CODE REVIEW

File: packages/cubejs-cubestore-driver/src/CubeStoreQueueDriver.ts [hunk 1]
```diff
@@ -0,0 +1,312 @@
...
+  QueueDriverOptions,
+  AddToQueueQuery,
+  AddToQueueOptions, AddToQueueResponse, QueryKey,
+} from '@cubejs-backend/base-driver';
+
+import crypto from 'crypto';
+import { CubeStoreDriver } from './CubeStoreDriver';
+
+function hashQueryKey(queryKey: QueryKey) {
+  return crypto.createHash('md5').update(JSON.stringify(queryKey)).digest('hex');
```
Comment (ovr) [inline packages/cubejs-cubestore-driver/src/CubeStoreQueueDriver.ts:16]: persistent keys will be in another PR.

File: packages/cubejs-query-orchestrator/src/orchestrator/RedisPool.ts [hunk 2]
```diff
@@ -79,7 +79,7 @@ export class RedisPool {
     }
   }
 
-  public release(client) {
+  public release(client: AsyncRedisClient) {
     if (this.pool) {
       this.pool.release(client);
```
Comment (ovr) [inline packages/cubejs-query-orchestrator/src/orchestrator/RedisPool.ts:84]: this.pool.release is async

**Review by buntarb — APPROVED** [review]
